src/Solnet.Rpc/SolanaRpcBatchComposer.cs

```diff
@@ -0,0 +1,384 @@
+        {
+            var reqs = this.CreateJsonRequests();
+            var response = await client.SendBatchRequestAsync(reqs);
+            if (response.WasSuccessful)
+                return ProcessBatchResponse(response);
+            else
+                throw new ApplicationException($"Batch was unsuccessful: {response.Reason}");
+        }
+
+        /// <summary>
+        /// Handles the conversion of the generic JSON deserialized response objects to the native types.
+        /// </summary>
+        /// <param name="response"></param>
+        /// <returns></returns>
+        public JsonRpcBatchResponse ProcessBatchResponse(RequestResult<JsonRpcBatchResponse> response)
+        {
+            // TODO - make sure result was successful
+
+            // sanity check response matches request
+            if (_reqs.Count != response.Result.Count) throw new ApplicationException($"Batch req/resp size mismatch {_reqs.Count}/{response.Result.Count}");
+
+            // transfer expected type info to individual
+            // batch response items
+            for (int ix = 0; ix < _reqs.Count; ix++)
+            {
+                var req = _reqs[ix];
+                var resp = response.Result[ix];
+
+                // set the runtime type
+                resp.ResultType = req.ResultType;
+
+                // catch any type mapping exceptions and feed into callback
+                bool callbackInvoked = false;
+                try
+                {
+                    // translate generic JSON deserialized content into POCO runtime types
+                    if (req.ResultType != null)
+                    {
+                        resp.Result = MapJsonTypeToNativeType(resp.Result, req.ResultType);
+                    }
+
+                }
+                catch (Exception ex)
+                {
+
+                    // invoke callback with the exception
+                    if (req.Callback != null)
+                    {
+                        callbackInvoked = true;
+                        req.Callback.Invoke(null, ex);
+                    }
+
+                }
+
+                // invoke callback with safe mapped data type
+                if (!callbackInvoked && req.Callback != null)
+                {
+                    callbackInvoked = true;
+                    req.Callback.Invoke(resp, null);
+                }
+
+            }
+
+            // reset ready for reuse
+            Clear();
+            
+            // pass back the JSON batch innards
+            return response.Result;
+        }
+
+        /// <summary>
+        /// Process a failed batch response
+        /// </summary>
+        /// <param name="response">The failed batch RPC response</param>
+        /// <returns></returns>
+        public JsonRpcBatchResponse ProcessBatchFailure(RequestResult<JsonRpcBatchResponse> response)
+        {
+            // TODO - decide best approach - invoke each callback with exception?
```

Comment:
I assume there is some error handling missing here?

Reply:
Comment removed - implemented notification of batch failure to all callbacks of batch failure via `BatchRequestException` exception.